Automattic/woocommerce-payments
# Gutenberg Block Widget for Currency Switcher

Fixes #1767 

#### Changes proposed in this Pull Request









- Implements a block widget for the Multi-Currency currency switcher.

#### Testing instructions





As a block widget:
* Go to `Appearance -> Widgets` in WP Admin, add a new block to the sidebar by clicking the plus icon at the bottom of the existing sidebar widgets, and selecting `Currency Switcher`. (note: you may see a warning at the top of this page, this is unrelated to this PR.)
* Click update on the top right corner of the page, then open a new tab on your site and verify that you can see the currency switcher in the sidebar. 
* Go back to `Appearance -> Widgets` and open the settings pane on the right. Play with toggling the various settings on/off and verify that the preview of the widget updates accordingly (note that some settings are not implemented yet, see the TODOs at the bottom of this PR). After making some changes, click update, then go to a page on your site and verify the currency switcher style updated accordingly.
* Test changing the currency via the new widget that was added, and do a few quick smoke tests to make sure currency is converted as expected on the store and when purchasing a product.

As a block via the editor:
* Go to `Pages -> Add New` and add a test page to your site. Open the block editor for the new page. Add a new block to this page and select `Currency Switcher`. 
* Switch to Preview mode (or save the change then go to the page on your store) and verify that you can see the currency switcher on the page where you added it.
* Repeat the tests above (changing the options and ensuring the changes get applied) to make sure all is working as expected when the widget is added as a block on a page.

#### Remaining TODOs and Questions

- I renamed the current currency widget to 'Legacy Currency Switcher' because otherwise we'd have two widgets in the list called 'Currency Switcher' which may be confusing. Is this okay? How should we handle this going forwards?
- Currently, the preview (displayed in `Appearance -> Widgets` and in the Editor) is using a hardcoded list of test currencies. I tried to add [logic](https://github.com/Automattic/woocommerce-payments/blob/ceece0a65ab6c67593522de70284122efcba4579/client/multi-currency/blocks/currency-switcher.js#L277-L310) to load the enabled currencies here, but it didn't seem to work and I'm not sure why yet. It's not hugely important, because when the widget is actually used it'll use the real enabled currencies, this is just for the Editor Preview.
- Some of the styling options such as border, border radius and border color are currently not doing anything. While the settings are in place and are getting passed between front and backend correctly, I haven't yet figured out how to get the border to display nicely in line with the [designs](https://wcpay.wordpress.com/2021/04/07/front-end-currency-switchers-for-multi-currency/). Other settings like font size, font color and background color are working correctly.
- Note: as we always need the to show the latest enabled currencies, we needed to implement a dynamic block here. This means that while the frontend React code controls both the settings and the preview of the currency switcher, the actual live widget is rendered using the PHP code found in the `includes/MultiCurrency/CurrencySwitcherBlock` file. I mention this as it could be confusing, and any change to the PHP code which renders the live widget should also be reflected in the JS code which displays the editor preview of the widget.

-------------------


**Post merge**
